### PR TITLE
Implemented regularizer (rebased)

### PR DESCRIPTION
**Pull request by shinaoka**
OK, rebased.

* Add flag for warning on condition numbers
* Add regularizer to fitting
* Add oversampling to fitting
* Add max_dim option to FiniteTempBasis.

### CODE REVIEW

File: src/irbasis3/basis.py [hunk 1]
```diff
@@ -152,7 +152,7 @@ class FiniteTempBasis:
         gl = basis.s * basis.v(2.5)
         giw = gl @ basis.uhat([1, 3, 5, 7])
     """
-    def __init__(self, kernel, statistics, beta, eps=None, sve_result=None, _mitigate_sampling_points=True):
+    def __init__(self, kernel, statistics, beta, eps=None, sve_result=None, _mitigate_sampling_points=True, max_size=None):
```
Comment (mwallerb) [inline src/irbasis3/basis.py:155]: Suggest adding `max_size` to `IRBasis` as well -- we should keep the interface of `IRBasis` and `FiniteTempBasis` consistent.

File: src/irbasis3/sampling.py [hunk 2]
```diff
@@ -3,6 +3,30 @@
...
+        return _double(sampling_points)
+    else:
+        raise RuntimeError("Invalid input.")
+
+def _mul_with_1d_array(X, y, axis):
+    """ Multiply numpy ndarray X with 1d array y along a given axis """
+    ss = X.ndim * [None]
+    ss[axis] = slice(None)
+    return X * y[tuple(ss)]
+
```
Comment (mwallerb) [inline src/irbasis3/sampling.py:29]: Suggest moving these functions to the bottom of the file.

File: src/irbasis3/sampling.py [hunk 3]
```diff
@@ -22,26 +46,45 @@ class SamplingBase:
      - `basis` : IR Basis instance
      - `matrix` : Evaluation matrix is decomposed form
      - `sampling_points` : Set of sampling points
+     - `log_oversampling` : Oversampling factor. The number of the resultant sampling points will be increased approximately by 2^log_oversampling
```
Comment (mwallerb) [inline src/irbasis3/sampling.py:49]: I find `log_oversampling` confusing to use ... I think an oversampling factor would be clearer here (we can take `np.log2` later)

File: src/irbasis3/sampling.py [hunk 4]
```diff
@@ -22,26 +46,45 @@ class SamplingBase:
...
+     - `warn_cond` : Warn if the condition number is large.
+     - `regularizer` : None (disable), True (singular values), or a 1D array-like object of floats.
```
Comment (mwallerb) [inline src/irbasis3/sampling.py:51]: I am unsure of the value of this ... why do we need to regularize here? The matrices seem to be conditioned well-enough?

File: src/irbasis3/sampling.py [hunk 5]
```diff
@@ -22,26 +46,45 @@ class SamplingBase:
...
+        elif regularizer is True:
+            self.regularizer = basis.s
+        else:
+            self.regularizer = np.asarray(regularizer)
+
         self.basis = basis
         self.matrix = DecomposedMatrix(
-                        self.__class__.eval_matrix(basis, self.sampling_points))
+                        self.__class__.eval_matrix(basis, sampling_points) * self.regularizer[None,:]
+                        )
```
Comment (mwallerb) [inline src/irbasis3/sampling.py:69]: Suggest moving this to a separate class

```
class RegularizedMatrix(DecomposedMatrix):
    def __init__(self, a, regularizer, svd_result=None):
        super().__init__(a * regularizer, svd_result)
        self.regularizer = regularizer

    def __matmul__(self, x):
        return super().__matmul__(x) / self.regularizer

    def _lstsq(self, x):
        ...
```

**Review by mwallerb — APPROVED** [review]
Looks okay!

File: src/irbasis3/sampling.py (hunk 4 above)
Comment (shinaoka) [reply src/irbasis3/sampling.py:51]: I found the terminology of `regularization` confusing.
This rather makes the fitting matrix `ill-posed` as we discussed before.

If we take regularize = True, we force the expansion coefficients in IR decay like singular values.
The user can say, "OK, my QMC data is fitted well AND the coefficients decay fast. The choice of wmax is OK".

**Comment by shinaoka** [comment]
I will make a new PR after the package name is changed.

**Pull request closed without merging**